### PR TITLE
Delete Old Interface Logs If the Sum Size of Those Logs Exceeds Reasonable Size

### PR DESCRIPTION
**Pull request by BingShearer**
Right now the logs roll at 512 kb, and this will delete the oldest if there are more than 100 of them. Prevents Interface log dir from getting as enormous as mine did.

**QA TEST**

Note: Must be willing to temporarily sacrifice 51 Mb of Hard Drive space. 

- Navigate to your log directory, APPDATA/LOCAL/High Fidelity/Interface/Logs on Windows. 

- Run this script (https://gist.githubusercontent.com/BingShearer/1ea56ab5ccddc4f6be9389d29fda3ee3/raw/51f91c4a7455e2b1e60771787c150122623ca24a/logFlood.js) to fill your logs.

- Watch as the oldest one gets picked off as the total size exceeds 51 Mb.

### CODE REVIEW

**Comment by zzmp** [comment]
Can the total log size be configurable?

**Comment by hifi-gustavo** [comment]
Build console output for commit: be0446f9a90f51a9e0272c8e7a4a43de314232bf available at: https://gist.github.com/dc1863a8a72d31e9001d0c873d7053c9<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8244-be0446f.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8244-be0446f.dmg

**Comment by BingShearer** [comment]
@zzmp It's a variable at the top of FileLogger.cpp right now. It could be made a menu option at a later time. Same goes for the max log size.

File: interface/src/FileLogger.cpp [hunk 1]
```diff
@@ -71,6 +76,23 @@ void FilePersistThread::rollFileIfNecessary(QFile& file, bool notifyListenersIfR
...
+		qint64 totalSizeOfDir = 0;
+		foreach (QFileInfo dirItm, filesInDir){
+			if (totalSizeOfDir < MAX_LOG_DIR_SIZE){
+				totalSizeOfDir += dirItm.size();
+			}
+			else {
+				QFile file(dirItm.filePath());
+				file.remove();
+			}
+		}
```
Comment (Atlante45) [inline interface/src/FileLogger.cpp]: Coding standard:
- tabs instead of 4 spaces
- `if (...) {`
- `} else {`

**Comment by Atlante45** [comment]
👍 That'd be nice if this could apply to servers logs as well.

**Comment by AndrewMeadows** [comment]
My advice would be to not worry about making it configurable at this time.  The vast majority of users will not care as long as the log files don't fill the hard drive.  If we hear testimony from someone who needs to tweak that number then we'll consider exposing it as a config option.

**Comment by BingShearer** [comment]
O woops. I guess that isn't three spaces. Hold on a sec.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9641ae8cd816eb57f28366f4b66c3416df049a58 available at: https://gist.github.com/dc1863a8a72d31e9001d0c873d7053c9<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8244-9641ae8.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8244-9641ae8.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4fb9d8801c1dc2d95ac05415685505ebd9ddf556 available at: https://gist.github.com/dc1863a8a72d31e9001d0c873d7053c9<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8244-4fb9d88.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8244-4fb9d88.dmg

**Comment by OmegaHeron** [comment]
In so far as ---- nothing to do with this PR --- goes... everything is different when you're dealing with server and 10GB "drives" for typical droplets.  That can evaporate far far far more quickly than my 8TB of desktop storage.  It's all a matter of perspective and, frankly, there's little good to be said for one size fits all solutions.  I'm far more excited to see log level config flags like +/-critical +/-warning +/-debug +/-verbose etc etc etc.  While it all seems so inconsequential in terms of multi-core blah blah blah CPUs at some undreamt of GHz rate + Terabytes of storage - there's elegance to not wasting cycles doing things most users will never even see vs having M/G/T-bytes of pedantic debug logs that might come in handy.

**Comment by kevinmthomasse** [comment]
testing

**Comment by kevinmthomasse** [comment]
## QA Test
**QA Tester**: KevinMThomas
**QA Pass**: **QA Pass** SUCCESS
**Additional Test Results + Observations**:  Ran ```https://gist.githubusercontent.com/BingShearer/1ea56ab5ccddc4f6be9389d29fda3ee3/raw/51f91c4a7455e2b1e60771787c150122623ca24a/logFlood.js```.  After two hours, old logs were picked off after each new log generated inside the ```Local\High Fidelity\Interface\Logs``` which were 512k each totaling 51Mb.

File: interface/src/FileLogger.cpp [hunk 2]
```diff
@@ -24,9 +24,14 @@
...
 static const QString DATETIME_FORMAT = "yyyy-MM-dd_hh.mm.ss";
 static const QString LOGS_DIRECTORY = "Logs";
+static const QString IPADDR_WILDCARD = "[0-9]*.[0-9]*.[0-9]*.[0-9]*";
+static const QString DATETIME_WILDCARD = "20[0-9][0-9]-[0,1][0-9]-[0-3][0-9]_[0-2][0-9].[0-6][0-9].[0-6][0-9]";
+static const QString FILENAME_WILDCARD = "hifi-log_" + IPADDR_WILDCARD + "_" + DATETIME_WILDCARD + ".txt";
 // Max log size is 512 KB. We send log files to our crash reporter, so we want to keep this relatively
 // small so it doesn't go over the 2MB zipped limit for all of the files we send.
 static const qint64 MAX_LOG_SIZE = 512 * 1024;
+// Max log files found in the log directory is 100.
+static const qint64 MAX_LOG_DIR_SIZE = 512 * 1024 * 100;
```
Comment (SamGondelman) [inline interface/src/FileLogger.cpp]: we should probably add a variable like MAX_NUM_LOGS or something and make this MAX_LOG_SIZE * MAX_NUM_LOGS.  that'll make this more easily configurable in The Future

File: interface/src/FileLogger.cpp [hunk 3]
```diff
@@ -71,9 +76,24 @@ void FilePersistThread::rollFileIfNecessary(QFile& file, bool notifyListenersIfR
...
+            if (totalSizeOfDir < MAX_LOG_DIR_SIZE){
+                totalSizeOfDir += dirItm.size();
+            } else {
+                QFile file(dirItm.filePath());
+                file.remove();
+            }
+        }
     }
 }
-
```
Comment (SamGondelman) [inline interface/src/FileLogger.cpp]: we should put this blank line back to separate the functions

**Comment by BingShearer** [comment]
@SamGondelman I can add the space back, but the point was to not fill the HD even if hypothetically there were only three big logs, so I think size is actually better than num

**Comment by SamGondelman** [comment]
@BingShearer I gotcha, yeah that's fine then.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9ee64d114b3bca06a5f5ac9cbd6e8c828705a27c available at: https://gist.github.com/dc1863a8a72d31e9001d0c873d7053c9<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8244-9ee64d1.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8244-9ee64d1.dmg

**Pull request merged**